File: src/main/java/org/janelia/saalfeldlab/paintera/serialization/WindowProperties.java
package org.janelia.saalfeldlab.paintera.serialization;

import java.util.Optional;

import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import javafx.beans.property.IntegerProperty;
import javafx.beans.property.SimpleIntegerProperty;
import javafx.beans.value.ObservableBooleanValue;

public final class WindowProperties
{
	private final IntegerProperty initialWidth = new SimpleIntegerProperty(800);

	private final IntegerProperty initialHeight = new SimpleIntegerProperty(600);

	public final IntegerProperty widthProperty = new SimpleIntegerProperty(initialWidth.get());

	public final IntegerProperty heightProperty = new SimpleIntegerProperty(initialWidth.get());

	public ObservableBooleanValue hasChanged = widthProperty
			.isNotEqualTo(initialWidth)
			.or(heightProperty.isNotEqualTo(initialHeight));

	public void clean()
	{
		initialWidth.set(widthProperty.get());
		initialHeight.set(heightProperty.get());
	}

	public void populate(JsonObject serializedWindowProperties)
	{
		Optional.ofNullable(serializedWindowProperties.get(WindowPropertiesSerializer.WIDTH_KEY)).map
				(JsonElement::getAsInt).ifPresent(
				widthProperty::set);
		Optional.ofNullable(serializedWindowProperties.get(WindowPropertiesSerializer.HEIGHT_KEY)).map
				(JsonElement::getAsInt).ifPresent(
				heightProperty::set);
		clean();
	}

}

File: src/main/kotlin/org/janelia/saalfeldlab/paintera/serialization/WindowProperties.kt
package org.janelia.saalfeldlab.paintera.serialization

import javafx.beans.property.BooleanProperty
import javafx.beans.property.IntegerProperty
import javafx.beans.property.SimpleBooleanProperty
import javafx.beans.property.SimpleIntegerProperty
import org.janelia.saalfeldlab.fx.extensions.nonnull


class WindowProperties {

	companion object {
		private const val DEFAULT_WIDTH = 800
		private const val DEFAULT_HEIGHT = 600
	}

	internal val widthProperty: IntegerProperty = SimpleIntegerProperty(DEFAULT_WIDTH)
	var width by widthProperty.nonnull()

	internal val heightProperty: IntegerProperty = SimpleIntegerProperty(DEFAULT_HEIGHT)
	var height by heightProperty.nonnull()

	internal val fullScreenProperty: BooleanProperty = SimpleBooleanProperty(false)
	var isFullScreen: Boolean by fullScreenProperty.nonnull()

}
